# Task 1: Implementing Backtracking

In [130]:
# 1. The objective function is f(x) = x[0]**2 + x[1]**4
# 2. The optimal solution is (0, 0) obviously.

import numpy as np
from numpy.linalg import norm
import matplotlib.pyplot as plt
from scipy.optimize import minimize_scalar

In [131]:
"""
@def   : objective function 
@param : x is vector 
@return: a scalar
"""
def objective_func(x):
    return (x[0] + 2*x[1] - 7)**2 + (2*x[0] + x[1] - 5)**2

In [132]:
"""
@def   : gradient of objective function 
@param : x is vector 
@return: a vector 
"""
def grad_objective_func(x):
    return np.array([10*x[0] + 8*x[1] - 34, 8*x[0] + 10*x[1] - 38])

In [133]:
"""
@def   : hessian of objective function 
@param : x is vector 
@return: a matrix 
"""
def hessian_func(x):
    return np.matrix([
        [10, 8], [8, 10]
    ])

In [134]:
# Define a backtracking function
def backtracking(f, grad, p, x, c1, ro):
    a = 1
    while f(x + a*p) > f(x) + c1*a*np.matmul(p,grad(x)):
        a = ro*a
    return a

# Task 2: Backtracking Steepest Descent

In [141]:
# The stopping criteria is selected to be 1e-9 for the gradient norm.
# The starting point is [5, 5] for this problem
# The optimal solution is [0., 0.].

x0 = np.array([0, 0])
x_opt = np.array([1., 3.])
tol = 1e-9
c1 = 1e-4
ro = 0.1

In [142]:
"""
@def   : exact steepest descent method
@param : x is vector 
@return: a vector 
"""
def backtracking_steepest_descent(x0):  
    x = x0           
    p = -grad_objective_func(x) 
    global count
    count = 0
    while norm(p) > 1e-9:                                       
        a = backtracking(objective_func, grad_objective_func, p, x, c1, ro) 
        x = x + a*p                   
        p = -grad_objective_func(x)         
        count += 1
    return x


x = backtracking_steepest_descent(x0)
print(x)

[1. 3.]


In [143]:
count

111

In [144]:
"""
@def   : exact steepest descent method
@param : x is vector 
@return: a vector 
"""
def exact_steepest_descent(x0):             
    x = x0                                  
    p = -grad_objective_func(x) 
    global count
    count = 0
    while norm(p) > 1e-9:                  
        def subproblem1d(alpha):            
            return objective_func(x + alpha * p)  
                                            
        res = minimize_scalar(subproblem1d) 
        x = x + res.x * p                   
        p = -grad_objective_func(x) 
        count += 1

    return x


x = exact_steepest_descent(x0)
print(x)

[1. 3.]


In [145]:
count

13

In [146]:
alpha = 1e-3
"""
@def   : inexact steepest descent method
@param : x is vector 
@return: a vector 
"""
def inexact_steepest_descent(x0):              
    x = x0                                   
    p = -grad_objective_func(x) 
    global count
    count = 0
    while norm(p) > 1e-9:                    
        x = x + alpha * p                    
        p = -grad_objective_func(x)          
        count += 1
    return x


x = inexact_steepest_descent(x0)
print(x)

[1. 3.]


In [147]:
count

10871

a) The running time is certainly more than exact steepest descent but much much less than inexact. 

b) With a rho = 0.5 the iterations are reduce then go back up with 0.9. 

# Task 3: Backtracking Newton's Method

In [178]:
"""
@def   : objective function 
@param : x is vector 
@return: a scalar
"""
def objective_func(x):
    return (1-x[0])**2 + 100*(x[1] - x[0]**2)**2

In [179]:
"""
@def   : gradient of objective function 
@param : x is vector 
@return: a vector 
This is a test
"""
def grad_objective_func(x):
    return np.array([-400*(x[1]-x[0]**2)*x[0]-2*(1-x[0]), 200*(x[1]-x[0]**2)])

In [180]:
def hessian_func(x):
    return np.matrix([
        [-400*(x[1]-3*x[0]**2)+2, -400*x[0]], [-400*x[0], 200]
    ])

In [181]:
c1 = 1e-4
x0 = np.array([1.2, 1.2])
x_opt = np.array([1., 1.])
ro = 0.1

In [182]:
"""
@def   : exact steepest descent method
@param : x is vector 
@return: a vector 
"""
def backtracking_steepest_descent(x0):  
    x = x0           
    p = -grad_objective_func(x) 
    global count
    count = 0
    while norm(p) > 1e-9:                                       
        a = backtracking(objective_func, grad_objective_func, p, x, c1, ro) 
        x = x + a*p                   
        p = -grad_objective_func(x)         
        count += 1
    return x


x = backtracking_steepest_descent(x0)
print(x)

[1. 1.]


In [183]:
count

264

In [184]:
"""
@def   : Newton's method
@param : x is vector 
@return: a vector 
"""
def newton_method(x0):
    x = x0
    p = -grad_objective_func(x)
    h = hessian_func(x)
    global count
    count = 0
    while norm(p) > 1e-9:
        newton_dir = np.linalg.solve(h, p)
        a = backtracking(objective_func, grad_objective_func, p, x, c1, ro)
        x = x + a*newton_dir
        p = -grad_objective_func(x)
        h = hessian_func(x)
        count += 1
    return x

x = newton_method(x0)
print(x)

[1. 1.]


In [185]:
count

25543

At (1.2,1.2), the backtracking steepest descent takes 264 iterations and the backtracking newtons method takes 25000.

In [186]:
c1 = 1e-4
x0 = np.array([-1.2, 1])
x_opt = np.array([1., 1.])
ro = 0.1

In [187]:
x = backtracking_steepest_descent(x0)
print(x)

[1. 1.]


In [188]:
count

946

In [189]:
x = newton_method(x0)
print(x)

[1. 1.]


In [190]:
count

28335

The backtracking steepest descent took about 3.5 times longer while the backtracking newton method took only 0.1 times longer.

# Task 4: Heavy Ball Method

In [204]:
c1 = 1e-4
x0 = np.array([1.2, 1.2])
x_opt = np.array([1., 1.])
ro = 0.1
a = 1e-3
B = 0.9

In [205]:
"""
@def   : exact steepest descent method
@param : x is vector 
@return: a vector 
"""
def heavyball(x0):  
    x = [x0]          
    p = -grad_objective_func(x[-1]) 
    global a
    x.append(x[-1] + a*p)
    global count
    count = 0
    while norm(p) > 1e-9:                                       
        a = backtracking(objective_func, grad_objective_func, p, x[-1], c1, ro) 
        x.append(x[-1] + a*p + B*(x[-1] - x[-2]))
        p = -grad_objective_func(x[-1])         
        count += 1
    return x[-1]


x = heavyball(x0)
print(x)

[1. 1.]


In [206]:
count

1553

In [207]:
c1 = 1e-4
x0 = np.array([-1.2, 1.])
x_opt = np.array([1., 1.])
ro = 0.1
a = 1e-3
B = 0.9

In [209]:
print(heavyball(x0),count)

[1. 1.] 1695
